File: problems/07_trees/0102_binary_tree_level_order_traversal/solution.dart
// Problem: 0102. Binary Tree Level Order Traversal
// Difficulty: Medium

class TreeNode {
  int val; TreeNode? left, right;
  TreeNode([this.val = 0, this.left, this.right]);
}

TreeNode? fromLevel(List<int?> a) {
  if (a.isEmpty) return null;
  final root = TreeNode(a[0]!);
  final q = <TreeNode>[root]; var qi = 0, i = 1;
  while (qi < q.length && i < a.length) {
    final n = q[qi++];
    if (i < a.length && a[i] != null) { n.left = TreeNode(a[i]!); q.add(n.left!); }
    i++;
    if (i < a.length && a[i] != null) { n.right = TreeNode(a[i]!); q.add(n.right!); }
    i++;
  }
  return root;
}

List<List<int>> levelOrder(TreeNode? root) {
  if (root == null) return [];
  final out = <List<int>>[];
  var q = <TreeNode>[root];
  while (q.isNotEmpty) {
    final layer = <int>[];
    final nx = <TreeNode>[];
    for (final n in q) {
      layer.add(n.val);
      if (n.left != null) nx.add(n.left!);
      if (n.right != null) nx.add(n.right!);
    }
    out.add(layer);
    q = nx;
  }
  return out;
}

bool _eq(List<List<int>> a, List<List<int>> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i].length != b[i].length) return false;
    for (var j = 0; j < a[i].length; j++) if (a[i][j] != b[i][j]) return false;
  }
  return true;
}

void main() {
  if (!_eq(levelOrder(fromLevel([3, 9, 20, null, null, 15, 7])), [[3], [9, 20], [15, 7]])) throw 'ex1';
  if (!_eq(levelOrder(fromLevel([1])), [[1]])) throw 'one';
  if (!_eq(levelOrder(fromLevel([])), [])) throw 'empty';
  if (!_eq(levelOrder(fromLevel([1, 2, 3, 4, 5, 6, 7])), [[1], [2, 3], [4, 5, 6, 7]])) throw 'full';
  print('All tests passed!');
}

File: problems/07_trees/0102_binary_tree_level_order_traversal/solution.py
"""
Problem: 0102. Binary Tree Level Order Traversal
Difficulty: Medium
URL: https://leetcode.com/problems/binary-tree-level-order-traversal/

解法：BFS 逐層收集，O(n)/O(w)。
"""

from __future__ import annotations

from typing import Optional, List

from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right

def from_level(arr: List[Optional[int]]) -> Optional[TreeNode]:
    if not arr: return None
    root = TreeNode(arr[0]); q = deque([root]); i = 1
    while q and i < len(arr):
        n = q.popleft()
        if i < len(arr) and arr[i] is not None: n.left = TreeNode(arr[i]); q.append(n.left)
        i += 1
        if i < len(arr) and arr[i] is not None: n.right = TreeNode(arr[i]); q.append(n.right)
        i += 1
    return root

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        pass


if __name__ == "__main__":
    solution = Solution()

    # Test 1
    # result = solution.solve(...)
    # assert result == expected

    print("All tests passed!")
